src/Uno.Toolkit.UI/Behaviors/ItemsRepeaterExtensions.IncrementalLoading.cs

```diff
@@ -0,0 +1,231 @@
+			{
+				ir.EffectiveViewportChanged += OnEffectiveViewportChanged;
+
+				SetIncrementalLoadingSubscription(ir, Disposable.Create(() => ir.EffectiveViewportChanged -= OnEffectiveViewportChanged));
+			}
+			else
+			{
+				GetIncrementalLoadingSubscription(ir)?.Dispose();
+			}
+		}
+
+		// The EffectiveViewport is the intersection of all known viewports that contain the ItemsRepeater in their sub-tree
+		private static async void OnEffectiveViewportChanged(FrameworkElement sender, EffectiveViewportChangedEventArgs args)
+		{
+			if (sender is not ItemsRepeater ir || ir.ItemsSource is not ISupportIncrementalLoading incrementalLoading) return;
+			if (GetIsLoading(ir)) return;
+
+			var orientation = GetOrientation(ir);
+
+			var averageItemLength = GetAverageItemLength(ir, orientation);
+
+			(double itemsRepeaterMajorSize, double viewPortLength, double viewportEdge) = orientation switch
+			{
+				Orientation.Vertical => (ir.ActualHeight, args.EffectiveViewport.Height, args.EffectiveViewport.Bottom),
+				Orientation.Horizontal => (ir.ActualWidth, args.EffectiveViewport.Width, args.EffectiveViewport.Right),
+				_ => throw new InvalidOperationException($"Invalid orientation value: {orientation}"),
+			};
+
+			// A "page" is defined as the estimated number of items that could fit in the ItemsRepeater's EffectiveViewport.
+			var pageSize = averageItemLength > 0 ? viewPortLength / averageItemLength : 0d;
+
+			var desiredItemBuffer = (GetIncrementalLoadingThreshold(ir) + 1) * pageSize;
+
+			var distanceToEnd = itemsRepeaterMajorSize - viewportEdge.FiniteOrDefault(0d);
+
+			var remainingItems = averageItemLength > 0 ? distanceToEnd / averageItemLength : 0d;
+
+			if (remainingItems <= desiredItemBuffer && incrementalLoading.HasMoreItems)
+			{
+				SetIsLoading(ir, true);
+				try
+				{
+					await incrementalLoading.LoadMoreItemsAsync((uint)(GetDataFetchSize(ir) * Math.Max(1, pageSize)));
+				}
+				catch (Exception e)
+				{
+					ir.Log().Error("An error occurred while loading more items", e);
+				}
+				finally
+				{
+					SetIsLoading(ir, false);
+				}
+			}
+		}
+
+		private static Orientation GetOrientation(ItemsRepeater ir)
+		{
+			if (TryGetOrientation(ir, out var orientation) && orientation is { } o)
+			{
+				return o;
+			}
+
+			var layout = ir.Layout;
+			var property = layout.FindDependencyProperty<Orientation>("OrientationProperty");
+			return property != null && layout.GetValue(property) is Orientation orien ? orien : Orientation.Vertical;
+		}
+
+		private static bool TryGetOrientation(ItemsRepeater ir, out Orientation? orientation)
+		{
+			(bool result, orientation) = ir.Layout switch
+			{
+				StackLayout sl => (true, sl.Orientation),
+				UniformGridLayout ugl => (true, ugl.Orientation),
+#if HAS_UNO
+				FlowLayout flow => (true, flow.Orientation),
+#endif
+				_ => (false, default)
+			};
+
+			return result;
+		}
+
+		private static double GetAverageItemLength(ItemsRepeater ir, Orientation orientation)
```

Comment:
couldve been cleaner:
```suggestion
		private static double GetAverageItemLength(ItemsRepeater ir, Orientation orientation) => ir.GetChildren()
			.Select(x => orientation == Orientation.Vertical ? fe.ActualHeight : fe.ActualWidth)
			.DefaultIfEmpty(0d)
			.Average();
```